From 4661dc5de3f9fa82972b5453d9ab8eb65ec5df01 Mon Sep 17 00:00:00 2001
From: Sam Leffler <sleffler@chromium.org>
Date: Thu, 29 Dec 2011 09:57:03 -0800
Subject: [PATCH] CHROMIUMOS: wpa_supplicant: notify bss signal changes

Update the current bss's signal strength in response to
EVENT_SIGNAL_CHANGE and generate a dbus signal (as needed).  This allows
clients to more closely track the current signal maintained for the
current bss.

BUG=chromium-os:17092
TEST=disable bgscan and then do a walk-around test to verify CQM events generate dbus PropertyChange signals

Change-Id: Ifa1c582f40e37d309ffb71a655b336bbd78c64a7
---
 wpa_supplicant/bss.c    | 9 +++++++++
 wpa_supplicant/bss.h    | 2 ++
 wpa_supplicant/events.c | 3 +++
 3 files changed, 14 insertions(+)

diff --git a/wpa_supplicant/bss.c b/wpa_supplicant/bss.c
index 4994e6e..c9e74e9 100644
--- a/wpa_supplicant/bss.c
+++ b/wpa_supplicant/bss.c
@@ -560,6 +560,15 @@ wpa_bss_update(struct wpa_supplicant *wpa_s, struct wpa_bss *bss,
 	return bss;
 }
 
+void wpa_bss_update_signal(struct wpa_supplicant *wpa_s,
+			   struct wpa_bss *bss, int cur_signal)
+{
+	if (bss->level != cur_signal) {
+		bss->level = cur_signal;
+		wpas_notify_bss_signal_changed(wpa_s, bss->id);
+	}
+}
+
 
 /**
  * wpa_bss_update_start - Start a BSS table update from scan results
diff --git a/wpa_supplicant/bss.h b/wpa_supplicant/bss.h
index 4a624c5..ddde1a4 100644
--- a/wpa_supplicant/bss.h
+++ b/wpa_supplicant/bss.h
@@ -102,6 +102,8 @@ void wpa_bss_update_scan_res(struct wpa_supplicant *wpa_s,
 			     struct os_reltime *fetch_time);
 void wpa_bss_update_end(struct wpa_supplicant *wpa_s, struct scan_info *info,
 			int new_scan);
+void wpa_bss_update_signal(struct wpa_supplicant *wpa_s,
+			   struct wpa_bss *bss, int cur_signal);
 int wpa_bss_init(struct wpa_supplicant *wpa_s);
 void wpa_bss_deinit(struct wpa_supplicant *wpa_s);
 void wpa_bss_flush(struct wpa_supplicant *wpa_s);
diff --git a/wpa_supplicant/events.c b/wpa_supplicant/events.c
index 47d39dd..786a1cf 100644
--- a/wpa_supplicant/events.c
+++ b/wpa_supplicant/events.c
@@ -3380,6 +3380,9 @@ void wpa_supplicant_event(void *ctx, enum wpa_event_type event,
 			data->signal_change.current_signal,
 			data->signal_change.current_noise,
 			data->signal_change.current_txrate);
+		if (wpa_s->current_bss)
+			wpa_bss_update_signal(wpa_s, wpa_s->current_bss,
+				data->signal_change.current_signal);
 		break;
 	case EVENT_INTERFACE_ENABLED:
 		wpa_dbg(wpa_s, MSG_DEBUG, "Interface was enabled");
-- 
2.2.0.rc0.207.ga3a616c

